File: gateway/adapters/slack.py
from __future__ import annotations

import logging
from typing import Callable

from .base import ChannelAdapter

log = logging.getLogger(__name__)


class SlackAdapter(ChannelAdapter):
    """Slack adapter using slack_bolt (socket mode)."""

    def __init__(self, bot_token: str, app_token: str):
        self._bot_token = bot_token
        self._app_token = app_token
        self._client = None

    async def start(self, on_message: Callable) -> None:
        from slack_bolt.async_app import AsyncApp
        from slack_bolt.adapter.socket_mode.async_handler import AsyncSocketModeHandler

        app = AsyncApp(token=self._bot_token)
        self._client = app.client

        @app.event("message")
        async def handle(event, say):
            if event.get("bot_id"):
                return
            peer_id = event.get("user")
            if not peer_id:
                return
            text = event.get("text", "")

            async def deliver(pid: str, response: str) -> None:
                await say(text=response, channel=event["channel"])

            await on_message("slack", peer_id, text, deliver)

        handler = AsyncSocketModeHandler(app, self._app_token)
        log.info("Slack adapter starting (socket mode)")
        await handler.start_async()

    async def send(self, peer_id: str, text: str) -> None:
        if self._client:
            await self._client.chat_postMessage(channel=peer_id, text=text)

File: src/gateway/adapters/slack.ts
import { ChannelAdapter } from "./base.ts";
import type { OnMessageFn } from "../router.ts";

export class SlackAdapter extends ChannelAdapter {
  private client: unknown = null;

  constructor(
    private readonly botToken: string,
    private readonly appToken: string,
  ) {
    super();
  }

  async start(onMessage: OnMessageFn): Promise<void> {
    const { App } = await import("@slack/bolt");
    const { SocketModeReceiver } = await import("@slack/bolt");

    const app = new App({
      token: this.botToken,
      appToken: this.appToken,
      socketMode: true,
    });

    this.client = app.client;

    app.message(async ({ event, say }: Record<string, unknown>) => {
      const ev = event as Record<string, unknown>;
      if (ev["bot_id"]) return;
      const peerId = ev["user"] as string | undefined;
      if (!peerId) return;
      const text = (ev["text"] as string | undefined) ?? "";
      const channel = ev["channel"] as string;

      await onMessage("slack", peerId, text, async (_pid: string, chunk: string) => {
        await (say as (opts: Record<string, unknown>) => Promise<void>)({ text: chunk, channel });
      });
    });

    console.log("Slack adapter starting (socket mode)");
    await app.start();
  }

  async send(peerId: string, text: string): Promise<void> {
    if (this.client) {
      const c = this.client as { chat: { postMessage(opts: Record<string, unknown>): Promise<unknown> } };
      await c.chat.postMessage({ channel: peerId, text });
    }
  }
}
